VoodooPS2Trackpad/VoodooPS2SynapticsTouchPad.cpp

```diff
@@ -1566,6 +1566,19 @@ void ApplePS2SynapticsTouchPad::sendTouchData() {
     inputEvent.contact_count = transducers_count;
     inputEvent.timestamp = timestamp;
 
+    // Send release event for leaving fingers
+    static int last_transducers_count = 0;
+    if (transducers_count < last_transducers_count) {
+        for (int i = transducers_count; i < last_transducers_count; i++) {
+            auto& transducer = inputEvent.transducers[i];
+            transducer.timestamp = timestamp;
+            transducer.isTransducerActive = 0;
+        }
+        inputEvent.contact_count = last_transducers_count;
+    }
+    last_transducers_count = transducers_count;
+    // Clean up other old finger data
+    memset(inputEvent.transducers + inputEvent.contact_count, 0, (VOODOO_INPUT_MAX_TRANSDUCERS - inputEvent.contact_count) * sizeof(VoodooInputTransducer));
```

Comment:
I see in VoodooInput that they for some reason have a +1 in loop bound: https://github.com/kprinssu/VoodooInput/blob/623942d85f4fbb517d81561976a49d4b5ed2aedb/VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp#L84
Maybe that's the reason? VoodooPS2 doesn't expect it to access any data beyond the point. I'll ask @kprinssu why they did it.

Reply:
It's for stylus. I changed to only read existing transducers when there is no stylus.